provider/pkg/provider/deployment_settings.go

```diff
@@ -0,0 +1,585 @@
+			var azure pulumiapi.AzureOIDCConfiguration
+
+			if azureInput["tenantId"].HasValue() {
+				azure.TenantID = getSecretOrStringValue(azureInput["tenantId"])
+			}
+			if azureInput["clientId"].HasValue() {
+				azure.ClientID = getSecretOrStringValue(azureInput["clientId"])
+			}
+			if azureInput["subscriptionId"].HasValue() {
+				azure.SubscriptionID = getSecretOrStringValue(azureInput["subscriptionId"])
+			}
+
+			oidc.Azure = &azure
+		}
+
+		oc.OIDC = &oidc
+	}
+
+	return &oc
+}
+
+func getSecretOrStringValue(prop resource.PropertyValue) string {
+	switch prop.V.(type) {
+	case *resource.Secret:
+		return prop.SecretValue().Element.StringValue()
+	default:
+		return prop.StringValue()
+	}
+}
+
+func (ds *PulumiServiceDeploymentSettingsResource) Diff(req *pulumirpc.DiffRequest) (*pulumirpc.DiffResponse, error) {
+	olds, err := plugin.UnmarshalProperties(req.GetOlds(), plugin.MarshalOptions{KeepUnknowns: false, SkipNulls: true})
+	if err != nil {
+		return nil, err
+	}
+
+	news, err := plugin.UnmarshalProperties(req.GetNews(), plugin.MarshalOptions{KeepUnknowns: true, SkipNulls: true})
+	if err != nil {
+		return nil, err
+	}
+
+	diffs := olds.Diff(news)
+	if diffs == nil {
+		return &pulumirpc.DiffResponse{
+			Changes: pulumirpc.DiffResponse_DIFF_NONE,
+		}, nil
+	}
+
+	replaces := make([]string, len(diffs.ChangedKeys()))
+	for i, k := range diffs.ChangedKeys() {
+		replaces[i] = string(k)
+	}
+
+	return &pulumirpc.DiffResponse{
+		Changes:             pulumirpc.DiffResponse_DIFF_SOME,
+		Replaces:            replaces,
+		DeleteBeforeReplace: true,
+	}, nil
+}
+
+func (ds *PulumiServiceDeploymentSettingsResource) Check(req *pulumirpc.CheckRequest) (*pulumirpc.CheckResponse, error) {
+	news, err := plugin.UnmarshalProperties(req.GetNews(), plugin.MarshalOptions{KeepUnknowns: true, SkipNulls: true})
+	if err != nil {
+		return nil, err
+	}
+
+	var failures []*pulumirpc.CheckFailure
+	for _, p := range []resource.PropertyKey{"organization", "project", "stack"} {
+		if !news[(p)].HasValue() {
+			failures = append(failures, &pulumirpc.CheckFailure{
+				Reason:   fmt.Sprintf("missing required property '%s'", p),
+				Property: string(p),
+			})
+		}
+	}
+
+	return &pulumirpc.CheckResponse{Inputs: req.News, Failures: failures}, nil
+}
+
+func (ds *PulumiServiceDeploymentSettingsResource) Configure(config PulumiServiceConfig) {}
+
+func (ds *PulumiServiceDeploymentSettingsResource) Read(req *pulumirpc.ReadRequest) (*pulumirpc.ReadResponse, error) {
+	ctx := context.Background()
+
+	var stack pulumiapi.StackName
+	if err := stack.FromID(req.Id); err != nil {
+		return nil, err
+	}
+	settings, err := ds.client.GetDeploymentSettings(ctx, stack)
+	if err != nil {
+		return nil, err
+	}
+
+	dsInput := PulumiServiceDeploymentSettingsInput{
+		Stack:              stack,
+		DeploymentSettings: *settings,
+	}
+
+	properties, err := plugin.MarshalProperties(
+		dsInput.ToPropertyMap(),
+		plugin.MarshalOptions{},
+	)
```

Comment:
One sort of unfortunate side-effect of the underlying API is that performing a `refresh` will hand back state that will always cause a diff due to the missing secret values. Not sure that I have a good suggestion for avoiding that.

Reply:
The only thing that comes to mind is by making some affordances in the API to return the secret values. I think for now this is fine, we can document it as a limitation for now.

Reply:
Added an issue to track: https://github.com/pulumi/pulumi-pulumiservice/issues/123